h5pt.go

```diff
@@ -84,36 +96,149 @@ func (t *Table) ReadPackets(start, nrecords int, data interface{}) error {
 // Append appends packets to the end of a packet table.
 func (t *Table) Append(data interface{}) error {
 	rv := reflect.Indirect(reflect.ValueOf(data))
+	rp := reflect.Indirect(reflect.ValueOf(&data))
 	rt := rv.Type()
-	c_nrecords := C.size_t(0)
+	c_nrecords := C.size_t(1)
 	c_data := unsafe.Pointer(nil)
 
 	switch rt.Kind() {
+	case reflect.Slice, reflect.Array:
+		for i := 0; i < rv.Len(); i++ {
+			if err := t.Append(rv.Index(i).Interface()); err != nil {
+				return err
+			}
+		}
+		return nil
 
-	case reflect.Array:
-		c_nrecords = C.size_t(rv.Len())
-		c_data = unsafe.Pointer(rv.UnsafeAddr())
+	case reflect.Struct:
+		ptr := C.malloc(1)
+		ptr_offset := C.size_t(0)
+
+		for i := 0; i < rv.NumField(); i++ {
+			var data_ptr unsafe.Pointer
+			var data_size C.size_t
+			f := rv.Field(i)
+			ft := string(f.Type().Name())
+
+			switch ft {
+			case "string":
+				c_string := C.CString(f.String())
+				data_ptr = unsafe.Pointer(c_string)
+				t.strings = append(t.strings, (*C.char)(data_ptr))
+				data_size = C.strlen(c_string) + 1
+			case "float32":
+				val := C.float(float32(f.Float()))
+				data_ptr = unsafe.Pointer(&val)
+				data_size = C.size_t(unsafe.Sizeof(val))
+			case "float64":
+				val := C.double(float64(f.Float()))
+				data_ptr = unsafe.Pointer(&val)
+				data_size = C.size_t(unsafe.Sizeof(val))
+			case "int8":
+				val := C.int8_t(int8(f.Int()))
+				data_ptr = unsafe.Pointer(&val)
+				data_size = C.size_t(unsafe.Sizeof(val))
+			case "int32":
+				val := C.int32_t(int32(f.Int()))
+				data_ptr = unsafe.Pointer(&val)
+				data_size = C.size_t(unsafe.Sizeof(val))
+			case "int64":
+				val := C.int64_t(int64(f.Int()))
+				data_ptr = unsafe.Pointer(&val)
+				data_size = C.size_t(unsafe.Sizeof(val))
+			case "uint8":
+				val := C.uint8_t(uint8(f.Uint()))
+				data_ptr = unsafe.Pointer(&val)
+				data_size = C.size_t(unsafe.Sizeof(val))
+			case "uint32":
+				val := C.uint32_t(uint32(f.Uint()))
+				data_ptr = unsafe.Pointer(&val)
+				data_size = C.size_t(unsafe.Sizeof(val))
+			case "uint64":
+				val := C.uint64_t(uint64(f.Uint()))
+				data_ptr = unsafe.Pointer(&val)
+				data_size = C.size_t(unsafe.Sizeof(val))
+			case "bool":
+				val := C.uchar(0)
+				if f.Bool() {
+					val = 1
+				}
+				data_ptr = unsafe.Pointer(&val)
+				data_size = C.size_t(1)
+			default:
+				return fmt.Errorf("hdf5: Could not append struct member %s "+
+					"to PacketTable", ft)
+			}
+
+			ptr = C.realloc(ptr, ptr_offset+data_size)
+			C.memset(unsafe.Pointer(uintptr(ptr)+uintptr(ptr_offset)), 0, data_size)
+			C.memcpy(unsafe.Pointer(uintptr(ptr)+uintptr(ptr_offset)), data_ptr, data_size)
+			ptr_offset += data_size
```

Comment:
this doesn't take into account C memory layout+alignment.
ie: each platform has its own memory alignment rules (mainly depending on 32b/64b)

when we pass a `void*` to hdf5, pointing at a C-struct, hdf5 expects that C-struct to honour the C alignment rules.

```c
struct S {
    uint32_t f1; // -> offset=0
    uint16_t f2; // -> offset=4
    uint8_t  f3; // -> offset=6
    uint16_t f4; // -> offset=8
};
```

this will have `sizeof(S) == 12` (on my 64b machine) and not 9.